File: java/Hot100/Heap/Heap.java
package Hot100.Heap;

public class Heap {
    private void minHeapify(int[] arr, int i, int heapSize) {
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        int smallest = i;

        if (left < heapSize && arr[left] < arr[smallest]) {
            smallest = left;
        }
        if (right < heapSize && arr[right] < arr[smallest]) {
            smallest = right;
        }

        if (smallest != i) {
            swap(arr,i,smallest);
            minHeapify(arr, smallest, heapSize);
        }
    }

    // 建最小堆
    public void buildMinHeap(int[] arr) {
        int n = arr.length;
        for (int i = n / 2 - 1; i >= 0; i--) {
            minHeapify(arr, i, n);
        }
    }

    private void swap(int[]arr,int i,int j){
        int temp=arr[i];
        arr[i]=arr[j];
        arr[j]=temp;
    }

    // 堆排序（升序）
    public void heapSort(int[] arr) {
        int n=arr.length;
        // 不断取出堆顶最小值，放到数组末尾
        for (int i = n - 1; i > 0; i--) {
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;

            minHeapify(arr, 0, i);
        }
    }

    // 测试
    public static void main(String[] args) {
        int[] arr = {3, 9, 2, 1, 4, 5};
        Heap heap = new Heap();

        heap.buildMinHeap(arr);
        for (int num : arr) {
            System.out.print(num + " ");
        }

        // 输出： [9, 5, 4, 3, 2, 1]
        System.out.println();
        heap.heapSort(arr);
        for (int num : arr) {
            System.out.print(num + " ");
        }
    }
}

File: python/heap/heap.py
class Heap:
    def min_heapify(self, arr: list[int], i: int, n: int) -> None:
        smallest = i
        left = 2 * i + 1
        right = 2 * i + 2

        if left < n and arr[left] < arr[smallest]:
            smallest = left

        if right < n and arr[right] < arr[smallest]:
            smallest = right

        if smallest != i:
            arr[i], arr[smallest] = arr[smallest], arr[i]
            self.min_heapify(arr, smallest, n)

    def build_min_heap(self, arr: list[int]) -> None:
        n = len(arr)

        for i in range(n // 2 - 1, -1, -1):
            self.min_heapify(arr, i, n)

    def heap_sort(self, arr: list[int]) -> None:
        """
        Sorts the array in the descending order.
        """
        n = len(arr)
        self.build_min_heap(arr)

        for i in range(n - 1, 0, -1):
            arr[0], arr[i] = arr[i], arr[0]
            self.min_heapify(arr, 0, i)